databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultFanout.java

```diff
@@ -128,72 +130,80 @@ protected void shutDown() throws Exception {
+            List<EventData> rawEvents = _eventSource.get(1000);
+            peekTime.stop();
+
+            // If no events, sleep a little while before doing any more work to allow new events to arrive.
+            if (rawEvents.isEmpty()) {
+                // Update the lag metrics to indicate there is no lag
+                updateLagMetrics(null);
+                return false;
+            }
 
-        // Last chance to check that we are the leader before doing anything that would be bad if we aren't.
-        return isRunning() && copyEvents(rawEvents);
+            // Last chance to check that we are the leader before doing anything that would be bad if we aren't.
+            return isRunning() && copyEvents(rawEvents);
+        }
     }
 
     @VisibleForTesting
     boolean copyEvents(List<EventData> rawEvents) {
         // Read the list of subscriptions *after* reading events from the event store to avoid race conditions with
         // creating a new subscription.
+        final Timer.Context subTime = _metricRegistry.timer(metricName("fetch-subscriptions")).time();
         Iterable<OwnedSubscription> subscriptions = _subscriptionsSupplier.get();
+        subTime.stop();
 
         // Copy the events to all the destination channels.
         List<String> eventKeys = Lists.newArrayListWithCapacity(rawEvents.size());
         ListMultimap<String, ByteBuffer> eventsByChannel = ArrayListMultimap.create();
         SubscriptionEvaluator.MatchEventData lastMatchEventData = null;
         int numOutboundReplicationEvents = 0;
-        for (EventData rawEvent : rawEvents) {
-            eventKeys.add(rawEvent.getId());
+        try (Timer.Context ignored = _metricRegistry.timer(metricName("fanout")).time()) {
+            for (EventData rawEvent : rawEvents) {
+                eventKeys.add(rawEvent.getId());
 
-            ByteBuffer eventData = rawEvent.getData();
+                ByteBuffer eventData = rawEvent.getData();
 
-            SubscriptionEvaluator.MatchEventData matchEventData;
-            try {
-                matchEventData = _subscriptionEvaluator.getMatchEventData(eventData);
-            } catch (UnknownTableException e) {
-                continue;
-            }
+                SubscriptionEvaluator.MatchEventData matchEventData;
+                try {
+                    matchEventData = _subscriptionEvaluator.getMatchEventData(eventData);
+                } catch (UnknownTableException e) {
+                    continue;
+                }
 
-            // Copy to subscriptions in the current data center.
-            for (OwnedSubscription subscription : _subscriptionEvaluator.matches(subscriptions, matchEventData)) {
-                eventsByChannel.put(subscription.getName(), eventData);
-            }
+                // Copy to subscriptions in the current data center.
+                for (OwnedSubscription subscription : _subscriptionEvaluator.matches(subscriptions, matchEventData)) {
```

Comment:
I think it would also be useful to have a timer metric around the actual subscription evaluation versus flush time.  It evaluates up subscriptions until `FLUSH_EVENTS_THRESHOLD` events have accumulated, then flushes them.  With the current metrics I don't think we could distinguish time calling `subscriptionEvaluator.matches()` from calls to `flush()`, and of the calls in this method not already timed those seem like the likeliest bottlenecks.